### PR TITLE
teck-programmer: init at 1.1.1

### PR DESCRIPTION
**Pull request by bbjubjub2494**
###### Motivation for this change
This program allows to flash the firmware of a specific line of keyboards. I have one of those that I'm repairing and I intend to flash it today or tomorrow.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lourkeur** [comment]
`nixpkgs-review`wise, `netlify-cli` won't build but it's probably not my code's fault

```
> node install.js

Trying to install "esbuild-linux-64" using npm
Failed to install "esbuild-linux-64" using npm: Command failed: npm install --loglevel=error --prefer-offline --no-audit --progress=false esbuild-linux-64@0.8.57
npm ERR! code ENOTCACHED
npm ERR! request to https://registry.npmjs.org/esbuild-linux-64 failed: cache mode is 'only-if-cached' but no cached response available.

npm ERR! A complete log of this run can be found in:
npm ERR!     /build/.npm/_logs/2021-04-01T21_50_36_492Z-debug.log

Trying to download "https://registry.npmjs.org/esbuild-linux-64/-/esbuild-linux-64-0.8.57.tgz"
Failed to download "https://registry.npmjs.org/esbuild-linux-64/-/esbuild-linux-64-0.8.57.tgz": getaddrinfo ENOTFOUND registry.npmjs.org
Install unsuccessful
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! esbuild@0.8.57 postinstall: `node install.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the esbuild@0.8.57 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /build/.npm/_logs/2021-04-01T21_50_36_690Z-debug.log

**Comment by lourkeur** [comment]
when installed it looks like it works
```
louis at hadronsson in ~/Téléchargements [1]
 ↪ teck-firmware-upgrade TrulyErgonomic_v4_00.hex
Found device in normal mode on bus 3, device 4.
Switching to program mode.
Rescanning for devices in one second.
Found device in program mode on bus 3, device 5.
Uploading firmware.
Uploading done.
Switching back to normal mode.
Done.

File: nixos/modules/programs/teck-programmer.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+with lib;
+
+let
+  cfg = config.programs.teck-programmer;
+
+in
+{
+  options.programs.teck-programmer = {
+    enable = mkEnableOption "whether to enable programming the firmware of TECK keyboards";
```
Comment (midchildan) [inline nixos/modules/programs/teck-programmer.nix:11]:
```suggestion
    enable = mkEnableOption "user access to TECK keyboard firmware";
```

`mkEnableOption` generates descriptions that start with "Whether to enable".
https://github.com/NixOS/nixpkgs/blob/b45bd598dcdbc9db50876ee0b01408abf97bc404/lib/options.nix#L97

**Review by midchildan — CHANGES_REQUESTED** [review]

**Review by midchildan — APPROVED** [review]

**Review by midchildan — APPROVED** [review]

**Comment by lourkeur** [comment]
refreshed node-packages because of the merge conflict

**Comment by lourkeur** [comment]
Result of `nixpkgs-review pr 118228` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>netlify-cli</li>
  </ul>
</details>
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>antora</li>
    <li>epgstation</li>
    <li>gtop</li>
    <li>iosevka</li>
    <li>joplin</li>
    <li>lumo</li>
    <li>pm2</li>
    <li>postcss-cli</li>
    <li>redoc-cli</li>
    <li>vimPlugins.coc-metals</li>
    <li>vimPlugins.coc-prettier</li>
    <li>vimPlugins.coc-pyright</li>
    <li>vimPlugins.coc-stylelint</li>
    <li>vscode-extensions.matklad.rust-analyzer</li>
    <li>vscode-extensions.ms-python.vscode-pylance</li>
    <li>vscode-extensions.vadimcn.vscode-lldb</li>
  </ul>
</details>

**Comment by lourkeur** [comment]
That should make things smoother on nixpkgs-review.

Result of `nixpkgs-review pr 118228` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>teck-programmer</li>
  </ul>
</details>

**Comment by lourkeur** [comment]
@midchildan can you review again? This should hopefully be the last time.

**Review by midchildan — APPROVED** [review]
LGTM

**Review by SuperSandro2000 — APPROVED** [review]
package LGTM

File: nixos/modules/programs/teck-programmer.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+    services.udev.packages = [
+      (pkgs.runCommand "teck-udev-rules" { } ''
+        mkdir -p $out/etc/udev/rules.d
+        ln -s -t $out/etc/udev/rules.d ${cfg.package}/lib/node_modules/teck-programmer/40-teck.rules
+      '')
+    ];
```
Comment (aanderse) [inline nixos/modules/programs/teck-programmer.nix:26]: I don't know much about `node` packaging in nixpkgs, but can we modify it so `40-teck.rules` is placed in the appropriate folder so all you need is `services.udev.packages = cfg.package;`?

File: nixos/modules/programs/teck-programmer.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+in
+{
+  options.programs.teck-programmer = {
+    enable = mkEnableOption "user access to TECK keyboard firmware";
+    package = mkOption {
+      type = types.package;
+      default = pkgs.teck-programmer;
+      description = "the teck-programmer package";
+    };
```
Comment (aanderse) [inline nixos/modules/programs/teck-programmer.nix:16]: Is there any other option? Do we need this?

**Review by aanderse — COMMENTED** [review]
I have no idea if my comments are relevant or not... Let me know!

File: nixos/modules/programs/teck-programmer.nix (hunk 3 above)
Comment (lourkeur) [reply nixos/modules/programs/teck-programmer.nix:16]: Potentially: there exists [one in C++](https://github.com/m-ou-se/tek) and [one in Rust](https://github.com/zgtm/tek-programmer-rs).

File: nixos/modules/programs/teck-programmer.nix (hunk 2 above)
Comment (lourkeur) [reply nixos/modules/programs/teck-programmer.nix:26]: not really at the nodePackages stage. However, I'm thinking I should follow the same recipe as `nitrokey-udev-rules` here so I will rewrite this.

File: nixos/modules/programs/teck-programmer.nix (hunk 3 above)
Comment (aanderse) [reply nixos/modules/programs/teck-programmer.nix:16]: That sounds like a reason to keep the package option around :+1:

File: nixos/modules/programs/teck-programmer.nix (hunk 2 above)
Comment (aanderse) [reply nixos/modules/programs/teck-programmer.nix:26]: Thanks for the answer. I'm good either way.

File: pkgs/os-specific/linux/teck-udev-rules/default.nix [hunk 4]
```diff
@@ -0,0 +1,22 @@
+{ lib, stdenv, teck-programmer }:
+
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/teck-udev-rules/default.nix:2]:
```suggestion
```

Superfluous newline

File: pkgs/os-specific/linux/teck-udev-rules/default.nix [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  version = lib.getVersion teck-programmer;
+
+  inherit (teck-programmer) src;
+
+  dontBuild = true;
+
+  installPhase = ''
+    mkdir -p $out/etc/udev/rules.d
+    cp 40-teck.rules $out/etc/udev/rules.d
+  '';
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/teck-udev-rules/default.nix:15]:
```suggestion
  installPhase = ''
    runHook preInstall
    install 40-teck.rules -D -t $out/etc/udev/rules.d/
    runHook postInstall
    '';
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/teck-udev-rules/default.nix (hunk 4 above)
Comment (aanderse) [reply pkgs/os-specific/linux/teck-udev-rules/default.nix:2]: Funny... I think @SuperSandro2000 complains when they aren't there sometimes. Standardized code formatter please, right? 😆

**Review by aanderse — APPROVED** [review]
LGTM :+1:

**Comment by AndersonTorres** [comment]
A last question: do it have some number version?

File: pkgs/os-specific/linux/teck-udev-rules/default.nix (hunk 4 above)
Comment (AndersonTorres) [reply pkgs/os-specific/linux/teck-udev-rules/default.nix:2]: There were three blank lines in this file. One is sufficient.

**Review by AndersonTorres — APPROVED** [review]

**Pull request merged**